### PR TITLE
Refactor \OC\Memcache\Factory

### LINKED ISSUE #13227: Expose Low Latency memcache
https://github.com/owncloud/core/blob/19b7911d79fca5c88849498298add83bd825565f/lib/private/memcache/factory.php#L64-L75 has a low latence cache which only uses in-memory caches so that no remote hosts will get called.

IMHO it would make sense to expose this as well in the public interface? There are often reasons where it makes sense to store something in a local cache but it wouldn't make sense to store it in a remote cache. (for example because the lookup time might be too long or it is only specific to that instance etc…)

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
Caches divided up into two groups: distributed and local. 'Low latency' is an alias for local caches, while the standard `create()` call tries to get distributed caches first, then local caches.

New functions:
- `createLocal()/createDistributed()` - creates local/distributed memcache
- `getAvailableLocal()/getAvailableDistributed()` - gets an available backend class, or `null`

Fixes #13227 

cc @LukasReschke @DeepDiver1975 @MorrisJobke

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Works - tested with the app management and the speed gain is still measurable :+1:

**Comment by Xenopathic** [comment]
@MorrisJobke There's a speed gain? Wasn't expecting that - with one type of memcache enabled it the code should be identical to before, it's when multiple backends are available that it makes a difference. But still, performance is performance I guess!

**Comment by MorrisJobke** [comment]
@Xenopathic Yes and no. I just tested the speed gain, that is directly visible if there is a cache in place and detected ;) (to be exactly #13212)

File: lib/private/memcache/factory.php [hunk 1]
```diff
@@ -24,63 +41,101 @@ public function __construct($globalPrefix) {
 	}
 
 	/**
+	 * @param array $cacheList
+	 * @return string|null
+	 */
+	private function getAvailable($cacheList) {
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: Enforce the type?

File: lib/private/memcache/factory.php [hunk 2]
```diff
@@ -24,63 +41,101 @@ public function __construct($globalPrefix) {
...
 	 * @return bool
 	 */
 	public function isAvailable() {
-		return XCache::isAvailable() || APCu::isAvailable() || APC::isAvailable() || Redis::isAvailable() || Memcached::isAvailable();
+		return $this->getAvailableDistributed() || $this->getAvailableLocal();
 	}
 
 	/**
-	 * get a in-server cache instance, will return null if no backend is available
+	 * see createLocal()
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: `@see`

File: lib/private/memcache/factory.php [hunk 3]
```diff
@@ -24,63 +41,101 @@ public function __construct($globalPrefix) {
...
-			return new APC($prefix);
-		} else {
-			return null;
-		}
+		return $this->createLocal($prefix);
 	}
 
 	/**
-	 * check if there is a in-server backend available
+	 * see getAvailableLocal()
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: `@see`

File: lib/private/memcache/factory.php [hunk 4]
```diff
@@ -24,63 +41,101 @@ public function __construct($globalPrefix) {
...
-		return XCache::isAvailable() || APCu::isAvailable() || APC::isAvailable() || Redis::isAvailable() || Memcached::isAvailable();
+		return $this->getAvailableDistributed() || $this->getAvailableLocal();
 	}
 
 	/**
-	 * get a in-server cache instance, will return null if no backend is available
+	 * see createLocal()
 	 *
 	 * @param string $prefix
 	 * @return null|Cache
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: Will this actually ever return null?

**Comment by DeepDiver1975** [comment]
from my understanding we need to expose these operations in the OCP namespace as well.
I'll come up with an api naming tomorrow - generally speaking the impl looks good - THX @Xenopathic

File: lib/private/memcache/factory.php [hunk 5]
```diff
@@ -24,63 +41,101 @@ public function __construct($globalPrefix) {
...
+	 */
+	public function getAvailableLocal() {
+		return $this->getAvailable($this->localCaches);
+	}
+
+	/**
+	 * create a distributed cache instance
+	 *
+	 * @param string $prefix
+	 * @return \OC\Memcache\Cache
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: May return `void`?

Comment (Xenopathic) [inline lib/private/memcache/factory.php]: `void` is not a type, and apparently using `@return void` is bad practise (PHP really returns `null` if there are no `return`s).

File: lib/private/memcache/factory.php [hunk 6]
```diff
@@ -17,6 +17,23 @@ class Factory implements ICacheFactory {
...
+		'\\OC\\Memcache\\Memcached'
+	];
+
+	/**
+	 * @var array $localCaches
+	 */
+	private $localCaches = [
+		'\\OC\\Memcache\\APCu',
+		'\\OC\\Memcache\\APC',
+		'\\OC\\Memcache\\XCache'
```
Comment (nickvergessen) [inline lib/private/memcache/factory.php]: please trailing `,`

File: lib/private/memcache/factory.php (hunk 5 above)
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: Mhm… Anyways that PHPDoc is lieing that way. We need to adjust it.

File: lib/private/memcache/factory.php [hunk 7]
```diff
@@ -17,6 +17,23 @@ class Factory implements ICacheFactory {
 	private $globalPrefix;
 
 	/**
+	 * @var array $distributedCaches
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: maybe use string[] ?

File: lib/private/memcache/factory.php [hunk 8]
```diff
@@ -17,6 +17,23 @@ class Factory implements ICacheFactory {
...
+	 */
+	private $distributedCaches = [
+		'\\OC\\Memcache\\Redis',
+		'\\OC\\Memcache\\Memcached',
+	];
+
+	/**
+	 * @var array $localCaches
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: maybe use string[]?

File: lib/private/memcache/factory.php [hunk 9]
```diff
@@ -24,63 +41,99 @@ public function __construct($globalPrefix) {
...
 
 	/**
-	 * get a in-server cache instance, will return null if no backend is available
-	 *
+	 * @see \OC\Memcache\Factory::createLocal()
 	 * @param string $prefix
-	 * @return null|Cache
+	 * @return \OC\Memcache\Cache
 	 */
 	public function createLowLatency($prefix = '') {
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: Do we still need this, I think `createLocal` does already take care of this? – I'd either remove the function completely and replace all occurrences or if it is widely used add a deprecation note?

File: lib/private/memcache/factory.php [hunk 10]
```diff
@@ -24,63 +41,99 @@ public function __construct($globalPrefix) {
...
+	 * @return \OC\Memcache\Cache
+	 */
+	public function createLocal($prefix = '') {
+		if ($cache = $this->getAvailableLocal()) {
+			return new $cache($this->globalPrefix . '/' . $prefix);
+		}
+	}
+
+	/**
 	 * get a cache instance, or Null backend if no backend available
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: Add a note that this first uses the distributed caches and after that falls backs to local ones.

File: lib/private/memcache/factory.php [hunk 11]
```diff
@@ -24,63 +41,99 @@ public function __construct($globalPrefix) {
 	}
 
 	/**
+	 * @param array $cacheList
+	 * @return string
```
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: may theoretically return nothing as well? - add a `return '';` at the end of the function?

File: lib/private/memcache/factory.php (hunk 9 above)
Comment (DeepDiver1975) [inline lib/private/memcache/factory.php]: > Do we still need this, I think createLocal does already take care of this? – I'd either remove the function completely and replace all occurrences or if it is widely used add a deprecation note?

Depends on the pubic api I wanted to come up with ....

File: lib/private/memcache/factory.php (hunk 5 above)
Comment (LukasReschke) [inline lib/private/memcache/factory.php]: (currently static code analyzers have no chance to detect that you need to check the return type – thus we can't catch a lot of potential pitfalls)

**Comment by DeepDiver1975** [comment]
Regarding API:
we basically have two options:
- introduce a second method to ICacheFactory like
```php
interface ICacheFactory{
	public function create($prefix = '');
	public function createLocal($prefix = '');
}
````

- introduce a second parameter to the existing method like
```php
interface ICacheFactory{
	public function create($prefix = '', onlyLocal = false);
}
````

@LukasReschke @MorrisJobke @icewind1991 opinions?

**Comment by LukasReschke** [comment]
The first one looks more logical to me as it may lead easily lead to confusion otherwise. For example in the `remove` case `$onlyLocal` with a default of `false` may imply that it deletes values from the distributed and the remote cache.

Same is obviously valid for the `get` and other functions which effectively only will ask the distributed cache.

**Comment by DeepDiver1975** [comment]
> The first one looks more logical to me as it may lead easily lead to confusion otherwise. For example in the remove case $onlyLocal with a default of false may imply that it deletes values from the distributed and the local cache.
> 
> Same is obviously valid for the get and other functions which effectively only will ask the distributed cache.

@LukasReschke the cachfactors only has two methods: create and is available - remove and get are on the explicit cache implementations

**Comment by LukasReschke** [comment]
> the cachfactors only has two methods: create and is available - remove and get are on the explicit cache implementations

I know. Let try to put some code here so that my concern are more understandable :-)

When we take a look at the following snippet it is pretty much clear what it does, it clears the whole cache from values that have the prefix `foo`. Since there is no differentiation between the in-memory cache and distributed caches this will always clear all values.

```php
\OC::$server->getMemCacheFactory()->create('foo')->clear();
```

With the proposed approach No. 2 we would now have:
```php
// This will now clear all values from the remote memcaches only if one is available. 
// Though the parameter is called `onlyLocal` which is implicitly set to false in the interface
// so from a look at the interface once (at least me) would assume that this deletes values from
// the local AND the distributed cache (because `onlyLocal` false obviously does not exclude `local`
// itself) 
\OC::$server->getMemCacheFactory()->create('foo')->clear();

// This is actually what developers have to do to create the local memcache and not the distributed one
\OC::$server->getMemCacheFactory()->create('foo', true)->clear();
```

My take here is that this – at least for me – seems somewhat confusing and with two different functions it is actually more clean and verbose what happens here.

**Comment by DeepDiver1975** [comment]
I see your point!

**Comment by DeepDiver1975** [comment]
The code samples are not really valid - create() will always return a new instance and therefore a valid code snippet looks like this:

```php
$cache = \OC::$server->getMemCacheFactory()->create('foo');
$cache->set(...);
$cache->get(...);
$cache->clear(...);

**Comment by LukasReschke** [comment]
Sure, but the key point stays there:

```php
// Doing this will only clear the distributed cache
$cache = \OC::$server->getMemCacheFactory()->create('foo');
$cache->set(...);
$cache->get(...);
$cache->clear(...);
```

```php
// Doing this will only clear the local cache
$cache = \OC::$server->getMemCacheFactory()->create('foo', true);
$cache->set(...);
$cache->get(...);
$cache->clear(...);
```

If we rename `onlyLocal` to `useLocal` the implementation would make look cleaner to me. Then it is pretty much clear that to clean the local cache you need to pass `true` as well.

**Comment by DeepDiver1975** [comment]
hmmm - I'm honestly not sure what the best approach might be - we also have to honor this flag in the isAvailable() method.

Let's add the public api in 8.1 - no need to rush here currently

**Comment by LukasReschke** [comment]
For me `$onlyLocal` implies:

- with a value of `true: "I will only store data in the local cache and also only delete it from there"
  - it currently does exactly that
- with a value of `false:`: "I will store data in whatever cache is there and delete from whatever cache has the data available."
  - it currently will store the data in whatever cache is there but will only delete from the memcache, so there may be an inconsistency when developers don't think about passing the parameter.

**Comment by LukasReschke** [comment]
> Let's add the public api in 8.1 - no need to rush here currently

Care to create a ticket for that so we don't forget it? :-)

**Comment by DeepDiver1975** [comment]
> Care to create a ticket for that so we don't forget it? :-)

let's simply shift https://github.com/owncloud/core/issues/13227 to 8.1

File: lib/private/memcache/factory.php (hunk 9 above)
Comment (Xenopathic) [inline lib/private/memcache/factory.php]: @LukasReschke ah, good point, I forgot that this isn't in the public API so we can deprecate functions immediately :smile:

**Comment by Xenopathic** [comment]
The one thing about this implementation that I don't like is the lack of ability to create a local-only cache but have the automatic fallback to Null if one isn't available. I think it's important to expose the raw `createLocal()` and `createDistributed()` methods though. What about extending `create()` with a parameter to define whether to use distributed caches, local caches, or both (default)? I'm talking about the private factory, not the public API BTW.

File: lib/private/memcache/factory.php [hunk 12]
```diff
@@ -24,63 +41,100 @@ public function __construct($globalPrefix) {
...
-		} else {
-			return new Null($prefix);
+	public function create($prefix = '') {
+		if ($cache = $this->createDistributed($prefix)) {
+			return $cache;
 		}
+		if ($cache = $this->createLocal($prefix)) {
+			return $cache;
+		}
+		return new Null();
```
Comment (MorrisJobke) [inline lib/private/memcache/factory.php]: From scrutinizer:

The return type of return `new \OC\Memcache\Null();` (`OC\Memcache\Null`) is incompatible with the return type declared by the interface `OCP\ICacheFactory::create` of type `OCP\ICache`.

Comment (Xenopathic) [inline lib/private/memcache/factory.php]: Fixed in the latest commit

**Comment by bantu** [comment]
Please consider addressing #6074 and changing behaviour to require explicit cache configuration. When a lot of cache drivers are available, cache guessing might cost considerable time on it's own. cc @LukasReschke

**Comment by DeepDiver1975** [comment]
Thanks @bantu for bringing this up again. Let's address this with 8.1

**Comment by Xenopathic** [comment]
@bantu Reading that issue, I completely agree with you. @LukasReschke @DeepDiver1975 should we go for the current approach, or require an explicit configuration option?

**Comment by nickvergessen** [comment]
@Xenopathic as per above, this one should be sufficient for 8.0 and 8.1 should be configurable

**Comment by Xenopathic** [comment]
Can I suggest we move this to the 8.1 milestone? It's unnecessary for oC 8.0 (doesn't fix any bugs or improve performance significantly) and it'd be best to directly implement the explicit cache @bantu mentioned.

**Comment by DeepDiver1975** [comment]
> Can I suggest we move this to the 8.1 milestone?

full ack - thx @MorrisJobke for moving

**Comment by LukasReschke** [comment]
8.1 is for discussion now. - Feel free to rebase :smile:

**Comment by Xenopathic** [comment]
@LukasReschke @bantu @nickvergessen @DeepDiver1975 Should I refactor this PR to allow for a configurable memcache backend? What should happen if no explicit memcache backend is set - do the automatic detection thing we do at the moment?

**Comment by nickvergessen** [comment]
if none is set, use the array implementation?

**Comment by Xenopathic** [comment]
Eww, the array implementation isn't a real memcache though :laughing:

**Comment by bantu** [comment]
"Explicit cache configuration" means that if no cache is configured, no cache is used. No autodetection whatsoever.

**Comment by Xenopathic** [comment]
@bantu OK, sounds reasonable. I take it the same applies if the requested cache is not available?

**Comment by bantu** [comment]
@Xenopathic Good question. If the drivers for the configured cache is missing, we might want to throw a fatal error right away. Not sure what is supposed to happen when e.g. the memcached server is not available temporarily.

**Comment by Xenopathic** [comment]
Right, memcache backend configuration is now done in config.php, with the keys `memcache.local` and `memcache.distributed`. `memcache.distributed` defaults to the value of `memcache.local` if not set. If the requested cache is unavailable (or null), the factory will return a Null backend for all `create*()` methods, however `isAvailable()` will still be strict and only return true if a non-Null backend is available.

Also, unit tests!

**Comment by icewind1991** [comment]
Looks good :+1:

**Comment by Xenopathic** [comment]
Call for review please @PVince81 @LukasReschke @DeepDiver1975 @MorrisJobke

File: config/config.sample.php [hunk 13]
```diff
@@ -783,9 +783,32 @@
 
 
 /**
+ * Memory caching backend configuration
+ *
+ * Available cache backends:
+ * - \OC\Memcache\APC        Alernative PHP Cache backend
```
Comment (LukasReschke) [inline config/config.sample.php]: Alternative

**Comment by LukasReschke** [comment]
As discussed in IRC: Needs some more work.

**Comment by LukasReschke** [comment]
Okay. – As discussed :+1: - needs documentation though…

**Comment by LukasReschke** [comment]
Also needs to get added to https://github.com/owncloud/core/wiki/ownCloud-8.1-Features

**Pull request merged**